### DISCUSSION TITLE
Understanding xbsp-src bootstrap-update

### DISCUSSION THREAD

**Discussion by gc-user** [void-linux/void-packages#48701] [Q&A]
This is just a question to get a better understanding.

So, I just updated my local copy of void-packages via "git pull -r" and then wanted to update the bootstrap packages via "xbsp-src bootstrap-update".

But instead of a successful update as I usually get, I got this error message:

./xbps-src bootstrap-update
=> xbps-src: updating repositories for host (x86_64)...
[*] Updating repository `https://repo-default.voidlinux.org/current/bootstrap/x86_64-repodata' ...
[*] Updating repository `https://repo-default.voidlinux.org/current/x86_64-repodata' ...
[*] Updating repository `https://repo-default.voidlinux.org/current/nonfree/x86_64-repodata' ...
[*] Updating repository `https://repo-default.voidlinux.org/current/debug/x86_64-repodata' ...
[*] Updating repository `https://repo-default.voidlinux.org/current/multilib/bootstrap/x86_64-repodata' ...
[*] Updating repository `https://repo-default.voidlinux.org/current/multilib/x86_64-repodata' ...
[*] Updating repository `https://repo-default.voidlinux.org/current/multilib/nonfree/x86_64-repodata' ...
=> xbps-src: updating software in / masterdir...
libcrypto3-3.1.5_1 in transaction breaks installed pkg `openssl-3.1.4_2'
libssl3-3.1.5_1 in transaction breaks installed pkg `openssl-3.1.4_2'
Transaction aborted due to unresolved dependencies.
libcrypto3-3.1.5_1 in transaction breaks installed pkg `openssl-3.1.4_2'
libssl3-3.1.5_1 in transaction breaks installed pkg `openssl-3.1.4_2'
Transaction aborted due to unresolved dependencies.
=> ERROR: xbps-src: failed to update base-chroot!

So, I wonder, why does the update process not just update openssl? After all, it was "xbsp-src binary-bootstrap" that installed openssl in the first place - so it's not like I manually installed the package or anything (I wouldn't even know how to do that with xbps-src)... And: the main system updated openssl to v3.1.5 just fine.

Yeah, I could just delete the master directory and let "xbsp-src binary-bootstrap" create a new one (which I tested and works). But, as stated above, this is about getting a better understanding of xbps-src, in this case specifically "xbsp-src bootstrap-update".

- Is there a way to manually update a bootstrap package with xbps-src? E.g. to "fix" such a problem like above.
- Can I somehow check if - for some reason - the openssl bootstrap package is "on hold" and thus won't get updated?

**Comment by Duncaen** [comment]
Sounds like you've had outdated binary packages in your local repository.

**Reply by gc-user** [reply]
Thanks for your reply!

Shouldn't take the update process take care of that?

Meanwhile I did some more tests, as I suspected such myself, and temporarily moved all openssl / libssl / libcrypt packages out of /hostdir/binpkgs and /hostdir/repocache-x86_64. Also, just in case and to be sure "the system" couldn't find any "trace" of those packages, I renamed x86_64-repodata in /hostdir/binpkgs (didn't find such a file in /hostdir/repocache-x86_64).

After that, the update process worked as expected.

But still, I wonder why the update process wouldn't just go ahead and update the outdated packages from remote, not matter which version is stored locally - put there (at least in the repocache folder) by the update process during an earlier update... The process could at least ask if it should update those packages if it otherwise could or would not continue with updating other packages.

Well, anyway, I learned a new "trick" today, so that's a good thing. :-)

**Reply by Duncaen** [reply]
> But still, I wonder why the update process wouldn't just go ahead and update the outdated packages from remote, not matter which version is stored locally

Its just how repositories in xbps work, repositories have priorities, and the local repos in xbps-src are prioritized over remote repos. The first repo providing a given package name is chosen to provide the package.

There should probably be a way to remove outdated packages from the local repos, at the moment there is only `./xbps-src update-local` which rebuilds all outdated packages.

Generally this can be avoided by not building random packages on the master branch and instead use features branches, in this case `xbps-src` will register packages in `hostdir/binpkgs/<branchname>`, which will only be used to install packages from if you are on this branch wile `hostdir/binpkgs` will always be used with a higher priority than the remote repositories, even if you are on a different branch.

**Reply by gc-user** [reply]
Thanks for the explanations and your tips!
I'll look into the "features branches package building" thing, sounds interesting.

**Reply by gc-user** [reply]
> Its just how repositories in xbps work, repositories have priorities, and the local repos in xbps-src are prioritized over remote repos. The first repo providing a given package name is chosen to provide the package.

I recently tripped over this issue again...

This time, there was a newer version of the package in `repocache-x86_64`.

So, taking your above quote in account, I'm assuming that because `binpkgs` alphabetically comes before repocache-x86_64` the outdated version of the respective package was still chosen over the newer one?

But doesn't xbps-src check srcpkgs to see which is the current version and then check where it can that one?

**Reply by Duncaen** [reply]
`repocache-x86_64` is not a repository its the cache for all remote repositories, its the same as `/var/cache/xbps`, just that its stored outside the masterdir so it can be shared between multiple masterdirs.

It still chooses local repos over remote repos thats, thats why its using your local package, not the cache.

**Reply by gc-user** [reply]
I see, thanks for your reply / explanation. Good to know.
